### PR TITLE
Image support

### LINKED ISSUE #28: Image Event Support
When in a chat room and an image event occurs
Then display the image message.

Note: The cli version will not be able to display the image, should print some sort of fallback.

### LINKED ISSUE #34: Add Support for sending Image Messages in Swing GUI
When in a chat room and the user has an image file intended to be sent as a message,
Then allow the user to select the image and send it as a message.

### PR DESCRIPTION
**Pull request by lotusronin**
Fixes #28 #34 

Adds support for image events, displays the image inline in the chat room in the swing version of the app, and caches image downloads, maintaining a map of matrix url to local file path.

### CODE REVIEW

**Comment by lotusronin** [comment]
I should note that this PR isn't quite ready for merging, but I wanted to get it out and maybe get some early feedback and or testing from everyone else.

Some issues left to be addressed in this version is choosing a proper cache directory and behaving correctly if the cached file path isn't a valid path. I'm also considering adding support for image uploading on this PR.

**Comment by lotusronin** [comment]
Another issue is that currently image downloads are triggered in the UI thread, when the Chat Room makes the array of messages. This can cause a slowdown on chat room entry and/or when loading a lot of uncached messages. Future improvement could queue up downloads in the background and the UI should present a placeholder image/loading icon, and let the UI update dynamically

**Comment by lotusronin** [comment]
Added initial support for sending images to chat rooms

**Review by enoxone — APPROVED** [review]

File: serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/Database.kt [hunk 1]
```diff
@@ -38,4 +39,25 @@ object Database {
     fun deleteAllSessions() {
         this.db?.sessionDbQueries?.deleteAllSessions()
     }
+
+    fun getImageInCache(url: String): String? {
+        val cached_img = this.db?.sessionDbQueries?.selectCachedMedia(url) { mxcUrl: String, localPath: String ->
```
Comment (Limvot) [inline serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/Database.kt:44]: Unnecessary variable?

File: serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt [hunk 2]
```diff
@@ -84,6 +85,79 @@ class MatrixSession(val client: HttpClient, val access_token: String, var transa
...
     }
 
+    fun sendImageMessage(url: String, room_id: String): Outcome<String> {
+        try {
+            val result = runBlocking {
+                val img_f = File(url)
+                val image_data = img_f.readBytes()
+                val f_size = image_data.size
+                var ct = ContentType.Image.JPEG
+                val mimetype =
```
Comment (Limvot) [inline serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt:108]: this is a tad odd - maybe val (ct, mimetype) = if (...) { (ContentType.Image.PNG, "image/png") } else ...?

File: serif_swing/src/main/kotlin/serif_swing/App.kt [hunk 3]
```diff
@@ -119,63 +134,106 @@ class SwingChatRoom(val transition: (MatrixState, Boolean) -> Unit, val panel: J
...
         var seq_vert_groups = layout.createSequentialGroup()
-        for ((sender, message) in m.messages) {
-            val sender = JLabel("$sender:  ")
-
-            val message = JTextArea(message)
-            message.setEditable(false)
-            message.lineWrap = true
-            message.wrapStyleWord = true
+        for (msg in m.messages) {
+            val _sender = msg.sender
```
Comment (Limvot) [inline serif_swing/src/main/kotlin/serif_swing/App.kt:190]: Hmm, this shouldn't have an underscore, when did it get here?

**Review by Limvot — APPROVED** [review]
Looks good, pretty much the only changes are addressed in the audio branch or my next branch

File: serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/Database.kt (hunk 1 above)
Comment (lotusronin) [reply serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/Database.kt:44]: Correct, we could just return the expression, I should change that in the audio branch.

File: serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt (hunk 2 above)
Comment (lotusronin) [reply serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt:108]: That is cleaner, to be honest I just wasn't sure of the syntax of the paired return and I remembered to set the content type and mimetype at different times when originally writing. @enoxone also mentioned this to me. As before, I can also change this in the audio branch

File: serif_swing/src/main/kotlin/serif_swing/App.kt (hunk 3 above)
Comment (lotusronin) [reply serif_swing/src/main/kotlin/serif_swing/App.kt:190]: I added this to get rid of the shadowing warning regarding the text area down below.

**Pull request merged**